src/main.cpp

```diff
@@ -509,110 +521,177 @@ int main(int argc, char* argv[])
+        args.include_dirs.push_back(optarg);
         break;
       case '#':
-        include_files.push_back(optarg);
+        args.include_files.push_back(optarg);
         break;
       case 'l':
-        listing = true;
+        args.listing = true;
         break;
       case 'c':
-        cmd_str = optarg;
+        args.cmd_str = optarg;
         break;
       case '$':
-        usdt_file_activation = true;
+        args.usdt_file_activation = true;
         break;
       case 'u':
-        safe_mode = false;
+        args.safe_mode = false;
         break;
       case 'b':
         break;
       case 'h':
         usage();
-        return 0;
+        exit(0);
       case 'V':
         std::cout << "bpftrace " << BPFTRACE_VERSION << std::endl;
-        return 0;
+        exit(0);
       case 'k':
-        helper_check_level++;
-        if (helper_check_level >= 3)
+        args.helper_check_level++;
+        if (args.helper_check_level >= 3)
         {
           usage();
-          return 1;
+          exit(1);
         }
         break;
       default:
         usage();
-        return 1;
+        exit(1);
     }
   }
 
   if (argc == 1) {
     usage();
-    return 1;
+    exit(1);
   }
 
   if (bt_verbose && (bt_debug != DebugLevel::kNone))
   {
     // TODO: allow both
     LOG(ERROR) << "USAGE: Use either -v or -d.";
-    return 1;
+    exit(1);
   }
 
-  if (!cmd_str.empty() && !pid_str.empty())
+  if (!args.cmd_str.empty() && !args.pid_str.empty())
   {
     LOG(ERROR) << "USAGE: Cannot use both -c and -p.";
     usage();
-    return 1;
+    exit(1);
   }
 
+  // Difficult to serialize flex generated types
+  if (args.helper_check_level && args.build_mode == BuildMode::AHEAD_OF_TIME)
+  {
+    LOG(ERROR) << "Cannot use -k[k] with --aot";
+    exit(1);
+  }
+
+  if (args.listing)
+  {
+    // Expect zero or one positional arguments
+    if (optind == argc)
+    {
+      args.search = "*:*";
+    }
+    else if (optind == argc - 1)
+    {
+      args.search = argv[optind];
+      if (args.search == "*")
+      {
+        args.search = "*:*";
+      }
+      optind++;
+    }
+    else
+    {
+      usage();
+      exit(1);
+    }
+  }
+  else
+  {
+    // When not in listing mode, we expect to find the script through -e or filename
+    if (args.script.empty() && argv[optind] == nullptr)
+    {
+      LOG(ERROR) << "USAGE: filename or -e 'program' required.";
+      exit(1);
+    }
+
+    // If no script was specified with -e flag, then we expect to find a script file
+    if (args.script.empty())
+    {  
+      args.filename = argv[optind];
+      optind++;
+    }
+
+    // Load positional parameters before driver runs so positional
+    // parameters used inside attach point definitions can be resolved.
+    while (optind < argc)
+    {
+      args.params.push_back(argv[optind]);
+      optind++;
+    }
+  }
+
+  return args;
+}
+
+int main(int argc, char* argv[])
+{
+  int err;
+
+  const Args args = parse_args(argc, argv);
```

Comment:
Main parse_args function, which returns all the parsed arguments into a const struct.